### ISSUE TITLE
JSON:API Inclusion of related resources

### ISSUE THREAD

**Issue by Finhas** [emberjs/data#3454]
We have an api built following the JSON::API 1.0 specs; all the data seems to be loading correctly through ember-data and the JSONAPI Serilaizer and Adaptor (polymorphic, async) except that related resources are sending unique requests to our api. It seems as though ember-data is expecting related data to always be returned instead of having it as a api/user option. 

> An endpoint **MAY** return resources related to the primary data by default.

http://jsonapi.org/format/#fetching-includes

Basically, can ember-data handle requesting and loading the included data: 
http://localhost:3000/api/apps/:id/responsibilities**?include=party**

Instead of requesting:
localhost:3000/api/apps/:id/responsibilities
and
[localhost:3000/api/party/:ids_from_responsibilities]

Thanks

**Comment by wecc** [comment]
Does you API return both `data` and `links { related }` for your relationships?

Currently links have precedence over data so if both are present data will be fetched by links instead of local data. We're considering this a bug and it's tracked in #3380.

**Comment by Finhas** [comment]
I was reading that issue and I believe this is related but not the same. It seems at though @wwwdata api is returning the `included` information from his primary data. The issue he is having is true for us too; data is being retrieved by links instead of local data but that isn't our current issue.

In JSON:API,  using 'included' to return information from primary data is optional. Basically, 'http://example.com/posts/1' could return this:

````
{
    "data": [
        {
            "type": "posts",
            "id": "1",
            "attributes": {
                "title": "JSON API paints my bikeshed!"
            },
            "relationships": {
                "author": {
                    "links": {
                        "self": "http://example.com/posts/1/relationships/author",
                        "related": "http://example.com/posts/1/author"
                    },
                    "data": {
                        "type": "people",
                        "id": "9"
                    }
                }
            },
            "links": {
                "self": "http://example.com/posts/1"
            }
        }
    ]
}
```` 

and http://example.com/posts/1?included=author` would return

````
{
    "data": [
        {
            "type": "posts",
            "id": "1",
            "attributes": {
                "title": "JSON API paints my bikeshed!"
            },
            "relationships": {
                "author": {
                    "links": {
                        "self": "http://example.com/posts/1/relationships/author",
                        "related": "http://example.com/posts/1/author"
                    },
                    "data": {
                        "type": "people",
                        "id": "9"
                    }
                }
            },
            "links": {
                "self": "http://example.com/posts/1"
            }
        }
    ],
    "included": [
        {
            "type": "people",
            "id": "9",
            "attributes": {
                "first-name": "Dan",
                "last-name": "Gebhardt",
                "twitter": "dgeb"
            },
            "links": {
                "self": "http://example.com/people/9"
            }
        }
    ]
}
````

**Comment by wwwdata** [comment]
Hi,

I think you are exactly describing the same problem that I have. My backend does not send the `included` object per default. I use the query parameter that you described and add it via `store.query('posts', {include: 'comments'});`

But it doesn't matter if you have a query parameter ir not, ember-data does not handle both scenarios correctly.

**Comment by Finhas** [comment]
Ok, so it looks like this has already been a point of discussion #1576 but it doesn't seem like it has been solved yet #3281. 

Is the plan to have this work:
````return this.store.findRecord('post', params.post_id, { adapterOptions: { query: { 'includes': 'comments' } } });````

In the meantime, I am trying to implement ````findOneQuery```` but am having some issues with JSONAPISerializer.extract

**Comment by wecc** [comment]
@Finhas 
> In the meantime, I am trying to implement findOneQuery but am having some issues with JSONAPISerializer.extract

The `JSONAPISerializer` only supports the new Serializer API so if you use `normalizeResponse` instead of `extract` it should work out.

**Comment by Finhas** [comment]
Thanks @wecc 

Also had to change it to findRecord and remove the typeClass from the push since it's deprecated.
